File: src/11_Container_With_Most_Water/11_Container_With_Most_Water.c
#include <stdio.h>

int maxArea(int* height, int heightSize);

int main(void) {
  int height1[] = {1, 8, 6, 2, 5, 4, 8, 3, 7};
  printf("%d\n", maxArea(height1, 9));

  int height2[] = {1, 1};
  printf("%d\n", maxArea(height2, 2));
  return 0;
}

int maxArea(int* height, int heightSize) {
  int left = 0, right = heightSize - 1;
  int bottonSide = right - left;
  int side = height[left] > height[right] ? height[right] : height[left];
  int result = bottonSide * side;

  while (left < right) {
    bottonSide = right - left;
    side = height[left] > height[right] ? height[right] : height[left];
    int square = side * bottonSide;

    if (result < square) {
      result = square;
    }

    if (height[left] > height[right]) {
      right--;
    } else {
      left++;
    }
  }

  return result;
}

File: src/11_Container_With_Most_Water/11_Container_With_Most_Water.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        left = 0
        right = len(height) - 1
        result = right * min(height[left], height[right])

        while(left < right):
            square = (right - left) * min(height[left], height[right])

            if(square > result):
                result = square
            
            if(height[left] > height[right]):
                right -= 1
            else:
                left += 1

        return result
    

if(__name__ == "__main__"):
    s1 = Solution()
    print(f"result = {s1.maxArea([1,8,6,2,5,4,8,3,7])}")